service/history/queues/executable.go

```diff
@@ -824,7 +824,8 @@ func (e *CircuitBreakerExecutable) Execute() error {
 	doneCb, err := e.cb.Allow()
 	if err != nil {
 		metrics.CircuitBreakerExecutableBlocked.With(e.metricsHandler).Record(1)
-		return err
+		// Return resource a exhausted error to ensure that this task is retried less aggressively and does not go to the DLQ.
+		return fmt.Errorf("%w: %w", serviceerror.NewResourceExhausted(enums.RESOURCE_EXHAUSTED_CAUSE_SYSTEM_OVERLOADED, "circuit breaker rejection"), err)
```

Comment:
Debating whether to add a new cause here in the API.

Reply:
why wrap it around as string error? I'm not sure if the system is handle the error type using errors.Is/As utility.

The resource exhausted error scope should be namespace, and the error cause may need something new?

Reply:
Handling logic can handle wrapped resource exhausted errors.

And looks like NewResourceExhausted was not updated to take in resource exhausted scope. We need to set the scope to namespace. cc @prathyushpv

As long as it's namespace scoped, system overloaded cause makes sense to me.

Reply:
https://github.com/temporalio/api-go/pull/156#discussion_r1540121532
I tried to modify the constructor to add this field. At that time @cretz suggested not to change this function signature because it is part of public API. We could add a constructor with options. But I thought it was simpler to create the struct directly. We do that everywhere we create a new ResourceExhaustedError.

Reply:
The scope is namespace + destination though. Should we add a new scope?

Reply:
I don't think the task error handling logic care about destination here. 
It just needs a way to know the resource exhausted error is due to a user issue instead off a system issue, and can calculate no user in memory latency based on that.  

I will need to update the error handling logic there to check for the scope. Right now I think the logic checks for specific resource exhausted cause. Not urgent though.

Reply:
Sounds good.
I think for observability we'd rather have a different scope but other than that I don't think it matters.